### ISSUE TITLE
Gravity bits (Xlib).

### ISSUE THREAD

**Issue by elraymond** [Kalmat/PyWinCtl#55]
Just fyi, the gravity bits you copied from the python ewmh module are wrong. Struggled myself with that a while ago. The correct bits are

```
1<<8  (= 0b100000000) for presence of x
1<<9  presence of y
1<<10 presence of width
1<<11 presence of height
1<<12 source indicator, application
1<<13 source indicator, pager
```

Reference: https://specifications.freedesktop.org/wm-spec/wm-spec-1.3.html#idm45140362862448

**Comment by Kalmat** [comment]
I will apply these values. Thank you SO MUCH for your help!!!

**Comment by Kalmat** [comment]
Thank you SO MUCH for your help!!!

I will follow your advice. Can you please help me a little bit more?

Is the part of the gravity correct, so I should replace the flags part like this (short example for 'x' value only)?

        gravity_flags = gravity | 0b0000100000000000
        if x is None:
            x = 0
        else:
            gravity_flags = gravity_flags | (1 << 8)

Furthermore, should I somehowe indicate that the value is not present when it actually isn't, like this?

        gravity_flags = gravity | 0b0000100000000000
        if x is None:
            x = 0
            gravity_flags = gravity_flags & ~(1 << 8)
        else:
            gravity_flags = gravity_flags | (1 << 8)

About source indication, the specs say: "The bits 12 to 15 indicate the source (see [the section called “Source indication in requests”](https://specifications.freedesktop.org/wm-spec/latest/ar01s09.html#sourceindication)), so 0001 indicates the application and 0010 indicates a Pager or a Taskbar."

I am not used to bitwise operations at all, sorry, but this seems to be more like bits 14 and 15, or am I totally wrong? Despite the correct bit, should it be done like this?

        if userAction:   # userAction stands for "application" request (as oposed to "pager" request)
            gravity_flags = gravity_flags | (1 << 12)
            gravity_flags = gravity_flags & ~(1 << 13)
        else:
            gravity_flags = gravity_flags & ~(1 << 12)
            gravity_flags = gravity_flags | (1 << 13)

Thank you again!

**Comment by elraymond** [comment]
That's correct.

It's the second byte of the gravity_flags combo where individual bits are set to indicate the presence of certain geometry parameters and the source type. The first, lower byte holds the actual gravity.

So 1<<8 sets the first bit of the second byte, 1<<9 the second bit of the second byte and so forth up to the sixth bit, 1<<13. Bits 7 and 8 of the second byte are unused, is what the specification tells us. So the code you wrote does exactly how the specification says it works.

I stumbled across that issue when I used the ewmh module in a script to shuffle media player windows around on the desktop, and move/resize didn't work at all as expected. It took me a little time to figure out that the module itself sets those flags in  a wrong manner.
